### DISCUSSION TITLE
insertOnConflictUpdate not changing null value on nullable fields

### DISCUSSION THREAD

**Discussion by Enhed** [simolus3/drift#1414] [Q&A]
can its be fixed?

```dart
import 'package:moor/moor.dart';

class OperationCargos extends Table {
  TextColumn get id => text()();
  TextColumn get cargo => text()();
  TextColumn get weight => text().nullable()();
  TextColumn get volume => text().nullable()();
  TextColumn get length => text().nullable()();
  TextColumn get width => text().nullable()();
  TextColumn get height => text().nullable()();
  TextColumn get plan => text()();

  BoolColumn get changed => boolean().clientDefault(() => false)();

  @JsonKey('barcode_plan') TextColumn get barcodePlan => text()();
  @JsonKey('barcode_fact') TextColumn get barcodeFact => text().nullable()();

  @override
  Set<Column>? get primaryKey => { id };
}
```

```dart
@UseMoor(
  tables: [OperationCargos]
)
class AppDatabase extends _$AppDatabase {
  AppDatabase() : super(_openConnection());

  @override
  int get schemaVersion => 1;

  Future foo() async {
    final original = OperationCargo(
      id: 'some_id',
      barcodePlan: '',
      cargo: '',
      changed: false,
      plan: '',
      width: '1',
      height: '2',
      length: '3',
      volume: '4',
      weight: '5'
    );
    
    await into(operationCargos).insertOnConflictUpdate(original);
    
    final copy = original.copyWith(
        width: null,
        height: null,
        length: null,
        volume: null,
        weight: null
    );

    await into(operationCargos).insertOnConflictUpdate(copy);
  }
}
```

fields `width, height, length, volume, weight` not changes, stay as in original object

ps. without method `copyWith` will be same problem

**Comment by simolus3** [comment]
Can you post a relevant table definition and a way to reproduce the behavior you're describing?

**Reply by Enhed** [reply]
updated

**Reply by simolus3** [reply]
Right, so I think the problem here is actually related to `copyWith`, not the `insertOnConflictUpdate`. When you pass a `null` parameter to `copyWith`, the _old_ value will be used (so it is equivalent to just not passing the parameter at all). Values can't be set to `null` with `copyWith`. We can't distinguish between parameters that haven't been set and those that are absent in Dart.

To solve this, I would recommend to enable the `generate_values_in_copy_with` [builder option](https://moor.simonbinder.eu/docs/advanced-features/builder_options/). With that option, you can write this to make it explicit:

```dart
    final copy = original.copyWith(
        width: Value(null),
        height: Value(null),
        length: Value(null),
        volume: Value(null),
        weight: Value(null)
    );
```

**Reply by Enhed** [reply]
thank u
its not working: The argument type `Object' can't be assigned to the parameter type 'Value<String?>`

**Reply by simolus3** [reply]
Where do you get that error? After enabling the option, you'll have to use `Value` for all nullable columns.

**Reply by Enhed** [reply]
![image](https://user-images.githubusercontent.com/36981278/129721575-59bf0f05-59fd-4216-82d9-cf7de8a201eb.png)

**Reply by simolus3** [reply]
`weight: Value(weight ?? null)` would be the equivalent.

**Reply by Enhed** [reply]
![image](https://user-images.githubusercontent.com/36981278/129732245-3aa5260f-5082-49f2-b25f-bd13b085e405.png)

**Reply by simolus3** [reply]
Did you enable the option and re-run the build? Is `weight` even nullable?
